ext/opcache/Optimizer/zend_inference.c

```diff
@@ -3332,14 +3426,14 @@ static int zend_update_type_info(const zend_op_array *op_array,
 				COPY_SSA_OBJ_TYPE(ssa_ops[i].op1_use, ssa_ops[i].op1_def);
 			}
 			if (ssa_ops[i].result_def >= 0) {
-				tmp = MAY_BE_ANY | MAY_BE_ARRAY_KEY_ANY | MAY_BE_ARRAY_OF_ANY | MAY_BE_ARRAY_OF_REF;
-				if (opline->result_type == IS_TMP_VAR) {
-					/* can't be REF  because of ZVAL_COPY_DEREF() usage */
-					tmp |= MAY_BE_RC1 | MAY_BE_RCN;
-				} else {
-					tmp |= MAY_BE_REF | MAY_BE_RC1 | MAY_BE_RCN | MAY_BE_ERROR;
+				tmp = zend_fetch_prop_type_info(op_array, ssa, opline, i, &ce);
+				if (opline->result_type != IS_TMP_VAR) {
+					tmp |= MAY_BE_REF | MAY_BE_ERROR;
```

Comment:
MAY_BE_REF implies MAY_BE_ANY, so the type hinting information is lost.

Reply:
As FETCH_OBJ_R returns a TMP_VAR, the type information is present for the main case. I don't think that the type information is really useful for the other cases.